drivers: dma: stm32: take "stream offset" into account when DMAV1 + DMAMUX is used

erwango opened the pull request:
Depending on the series in use dma_config should take into account a stream offset for the channel in use.

Fixes #93617

erwango committed:
drivers: adc: stm32: Take into account "stream offset" in dma config

Depending on the series in use dma_config should take into account
a stream offset for the channel in use.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango commented:
Probably we need to review this whole offset handling and see why it can't be directly dealt from inside dma driver.

mathieuchopstm commented:
> Probably we need to review this whole offset handling and see why it can't be directly dealt from inside dma driver.

From 
https://github.com/zephyrproject-rtos/zephyr/blob/2a88cb50f57cf010fc7d60b7e686f960417d6fd0/include/zephyr/drivers/dma/dma_stm32.h#L18-L31

DMAv1 + DMAMUX ==> `STM32_DMA_STREAM_OFFSET = 1` (this is the HW configuration on STM32G0)

However, neither `dmamux_stm32.c` nor `dma_stm32_v1.c` use `STM32_DMA_STREAM_OFFSET`.

erwango committed:
drivers: dma: dma_stm32: irq handler fix in DMAMUX config case

IRQ handler should take into account STM32_DMA_STREAM_OFFSET
also when DMAMUX is present.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
drivers: dma: stm32u5: Clean up stream offset code

STM32_DMA_STREAM_OFFSET is defined as 0 in case "dma u5" is in use.
Clean up code relating to this define.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

mathieuchopstm commented on drivers/dma/dma_stm32.c:
```diff
@@ -105,8 +105,9 @@ static void dma_stm32_irq_handler(const struct device *dev, uint32_t id)
+
 #ifdef CONFIG_DMAMUX_STM32
-	callback_arg = stream->mux_channel;
+	callback_arg = stream->mux_channel + STM32_DMA_STREAM_OFFSET;
```
I think we should instead apply the offset here:
https://github.com/zephyrproject-rtos/zephyr/blob/0cc6d525b6f7a14d64c6fec78d25e66eaea83e27/drivers/dma/dma_stm32.c#L656-L659

```diff
#ifdef CONFIG_DMAMUX_STM32
 		/* Each further stream->mux_channel is fixed here */
- 		config->streams[i].mux_channel = i + config->offset;
+ 		config->streams[i].mux_channel = i + config->offset + STM32_DMA_STREAM_OFFSET;
#endif /* CONFIG_DMAMUX_STM32 */
```

mathieuchopstm reviewed (commented):
Otherwise LGTM (though possibly now we need to remove the `id + STM32_DMA_STREAM_OFFSET` in the few drivers that have it...)

etienne-lms reviewed (approved):
Looks consistent to me, FWIW.
